ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BaseHandler.java

```diff
@@ -0,0 +1,90 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ */
+package com.github.ambry.frontend.s3;
+
+import com.github.ambry.commons.Callback;
+import com.github.ambry.frontend.Operations;
+import com.github.ambry.rest.RequestPath;
+import com.github.ambry.rest.RestRequest;
+import com.github.ambry.rest.RestResponseChannel;
+import com.github.ambry.rest.RestServiceException;
+import com.github.ambry.router.ReadableStreamChannel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.rest.RestUtils.*;
+import static com.github.ambry.rest.RestUtils.InternalKeys.*;
+
+/**
+ * Base class for all S3 handle classes that provides crosscutting functionality such as
+ * logging and response header handling
+ *
+ * @param <R> the type of the result the object in callbacks. It is {@link Void} for requests that don't
+ *           require a response body, otherwise it is {@link ReadableStreamChannel}.
+ */
+abstract public class S3BaseHandler<R> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(S3BaseHandler.class);
+
+  /**
+   * Handles the S3 request and construct the response.
+   *
+   * @param restRequest the {@link RestRequest} that contains the request headers and body.
+   * @param restResponseChannel the {@link RestResponseChannel} that contains the response headers and body.
+   * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @throws RestServiceException exception when the processing fails
+   */
+  abstract protected void doHandle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      Callback<R> callback) throws RestServiceException;
+
+  /**
+   * Process the request and construct the response, internally the invocation is delegated
+   * to method {@link #doHandle(RestRequest, RestResponseChannel, Callback)}, which is
+   * implemented by subclasses.
+   *
+   * @param restRequest the {@link RestRequest} that contains the request headers and body.
+   * @param restResponseChannel the {@link RestResponseChannel} that contains the response headers and body.
+   * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @throws RestServiceException exception when the processing fails
+   */
+  public void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      Callback<R> callback) throws RestServiceException {
+    String path = ((RequestPath) restRequest.getArgs().get(REQUEST_PATH)).getOperationOrBlobId(true);
+    if (!path.startsWith(Operations.NAMED_BLOB)) {
+      throw new RuntimeException("S3 request handler can only handle named blob requests");
```

Comment:
Just a question. Would it be better to send `400 bad request` to user by completing the callback with `new RestServiceException("S3 request handler can only handle named blob requests", RestServiceErrorCode.BadRequest` ? I think `RuntimeException` translates to `500 internal server error` to user.

Reply:
My understanding is that this should never happen, i.e. we internally translate /s3 to /named and do other checks before we let the request proceed, so I chose 500. However, if there is a chance for client to construct to land here, then I agree that we should return 400.